File: 57-insert-interval/insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
      int st = newInterval[0];
      int en = newInterval[1];
      int i = 0;
      int n = intervals.length;

      List<int[]> res = new ArrayList<>();

      while(i < n && st > intervals[i][1]){
        res.add(intervals[i++]);
      }

      while(i < n && en >= intervals[i][0]){
        st = Math.min(st , intervals[i][0]);
        en = Math.max(en , intervals[i][1]);
        i++;
      }

      res.add(new int[]{st , en});

      while(i < n){
        res.add(intervals[i++]);
      }

      int j = 0;
      int[][] ans = new int[res.size()][2];
      for(int[] interval : res){
        ans[j++] = interval;
      }
      return ans;
    }
}

File: 57-insert-interval/insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        st,end = newInterval
        res = []
        n = len(intervals)
        i = 0

        while i < n and intervals[i][1] < st:
            res.append(intervals[i])
            i += 1

        while i < n and intervals[i][0] <= end:
            st = min(intervals[i][0] , st)
            end = max(intervals[i][1] , end)
            i += 1

        res.append([st, end])

        while i < n:
            res.append(intervals[i])
            i += 1
        
        return res
        